### DISCUSSION TITLE
io.UnsupportedOperation: fileno

### DISCUSSION THREAD

**Discussion by aleksyeyev1995** [pymodbus-dev/pymodbus#2176] [Help] [answered]
Hello!
I've changed pc with new one and my modbus serial script doesn't work now.
I'm work in Windows. There is an error from console:
![image](https://github.com/pymodbus-dev/pymodbus/assets/97503543/efaed9ae-177c-4250-899a-d73899344df7)
Can you give me an advice how to resolve this problem?

**Comment by janiversen** [comment]
Did you install pyserial ?

It seems you are running the sync client in an async environment, is there a reason for that ? it is very inefficient, and will likely cause problems.

We supply the simple client example as both sync and async.

fileno() is an integral part of pyserial, so you either do not have it installed or have an old version installed.

**Comment by chihpingkuo** [comment]
I ran into the same error on my windows pc
![image](https://github.com/pymodbus-dev/pymodbus/assets/6510834/d696e07f-e165-4886-adb4-65f71b4e051c)
I have installed 
pymodbus==3.6.8
pyserial==3.5

I believe the reason is that pyserial doesn't support fileno() on windows ?
https://pythonhosted.org/pyserial/pyserial_api.html#serial.Serial.fileno
![image](https://github.com/pymodbus-dev/pymodbus/assets/6510834/5d910a6c-7eb8-42f3-ace7-03136c3b9f1a)

**Reply by janiversen** [reply]
We run tests multiple times daily, this includes running on a windows machine with python 3.9-3.12, so it is a bit hard to believe that it does not work,

fileno() in pyserial works on windows, however we have often seen that the com port or usb in use, either do not exist or do not have the sufficient permissions.

**Comment by chihpingkuo** [comment]
on my windows 10 pc,
python 3.12.3 + pymodbus 3.6.8 => async serial client example will throw io.UnsupportedOperation: fileno
python 3.11.9 + pymodbus 3.6.8 => async serial client example runs successfully
I don't understand why, but I will stick with python 3.11 for now

**Reply by janiversen** [reply, marked as answer]
Interesting, so it's a python 3.12 issue, and since pyserial in reality isn't maintained it probably have not been properly tested.

Staying with 3.11 is a good temporary solution, there is an open pull request that solves the problem.
